src/plugins/profiling/server/routes/search_flamechart.ts

```diff
@@ -12,47 +12,54 @@ import { getRemoteRoutePaths } from '../../common';
 import { FlameGraph } from './flamegraph';
 import { newProjectTimeQuery } from './mappings';
 
-function getSampledTraceEventsIndex(
-  sampleSize: number,
-  sampleCountFromSmallestTable: number
-): [string, number, number] {
-  if (sampleCountFromSmallestTable === 0) {
-    // If this happens the returned estimatedSampleCount may be very wrong.
-    // Hardcode sampleCountFromSmallestTable to 1 else an estimatedSampleCount
-    // of zero is returned.
-    sampleCountFromSmallestTable = 1;
-  }
-
-  const sampleRates: number[] = [1000, 125, 25, 5, 1];
-
-  let tableName: string;
-  let sampleRate = 0;
-  let estimatedSampleCount = 0;
-
-  for (let i = 0; i < sampleRates.length; i++) {
-    sampleRate = sampleRates[i];
+export interface DownsampledEventsIndex {
+  name: string;
+  sampleRate: number;
+}
 
-    // Fractional floats have an inherent inaccuracy,
-    // see https://docs.python.org/3/tutorial/floatingpoint.htm) for details.
-    // Examples:
-    //   0.01 as float32: 0.009999999776482582092285156250 (< 0.01)
-    //   0.01 as float64: 0.010000000000000000208166817117 (> 0.01)
-    // But if "N / f >= 1", which is the case below as N is >= 1 and f is <= 1,
-    // we end up with the correct result (tested with a small Go program).
-    estimatedSampleCount = sampleCountFromSmallestTable * (sampleRates[0] / sampleRate);
+// Return the index that has between targetSampleSize..targetSampleSize*samplingFactor entries.
+// The starting point is the number of entries from the profiling-events-5pow<initialExp> index.
+//
+// More details on how the down-sampling works can be found at the write path
+//   https://github.com/elastic/prodfiler/blob/bdcc2711c6cd7e89d63b58a17329fb9fdbabe008/pf-elastic-collector/elastic.go
+export function getSampledTraceEventsIndex(
+  targetSampleSize: number,
+  sampleCountFromInitialExp: number,
+  initialExp: number
+): DownsampledEventsIndex {
+  const maxExp = 11;
+  const samplingFactor = 5;
+  const fullEventsIndex: DownsampledEventsIndex = { name: 'profiling-events', sampleRate: 1 };
+
+  if (sampleCountFromInitialExp === 0) {
+    // Take the shortcut to the full events index.
+    return fullEventsIndex;
+  }
 
-    if (estimatedSampleCount >= sampleSize) {
-      break;
+  if (sampleCountFromInitialExp >= samplingFactor * targetSampleSize) {
+    // Search in more down-sampled indexes.
+    for (let i = initialExp + 1; i <= maxExp; i++) {
+      sampleCountFromInitialExp /= samplingFactor;
+      if (sampleCountFromInitialExp < samplingFactor * targetSampleSize) {
+        return { name: 'profiling-events-5pow' + i, sampleRate: 1 / samplingFactor ** i };
```

Comment:
`profiling-events-pow5` should probably be a global constant?

Reply:
Just pushed a constant